# **Sistema de recomendaciòn para la compra de terreno**

In [ ]:
# Importar la biblioteca pandas para el manejo y análisis de datos
import pandas as pd

# Importar la función de similitud del coseno para calcular la similitud entre vectores
from sklearn.metrics.pairwise import cosine_similarity

# Importar MinMaxScaler para escalar características a un rango específico (por ejemplo, [0,1])
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder

# Importar la biblioteca numpy para operaciones numéricas y manejo de arreglos
import numpy as np

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


##IMPORTE MI BASE DATOS "BasedeDatosDefinitiva" para trabajar con ella##


##**Generaciòn de datos**

In [ ]:
# Simulación de datos de terrenos
df=pd.read_csv('/content/drive/MyDrive/basedatos/BasedeDatosDefinitivaID.csv')
df

,id,area,topografia,forma,cam_conti,agua,luz,alcantarillado,dist_pist(Km),tip_doc,zonificación,tip_suelo,largo(mts),ancho(mts),vist_pano,población,ciudad,precio,precio(mts2)
0,1,108,llanura,horizontal,camino_medio_estrecho,si,si,si,1,titulo,residencial,rocoso,18.00,6.00,no,32312,Chancay,30000,277.777778
1,2,112,llanura,vertical,via_publica,si,si,si,2,titulo,residencial,rocoso,14.00,8.00,no,29671,Hualmay,30000,285.714286
2,3,120,llanura,vertical,camino_medio_estrecho,si,si,si,1,titulo,residencial,rocoso,15.00,8.00,si,156326,Barranca,35000,291.666667
3,4,121,llanura,horizontal,camino_medio_estrecho,si,si,si,0,titulo,residencial,rocoso,11.27,10.75,no,37121,Huaura,37000,305.785124
4,5,125,llanura,vertical,camino_medio_estrecho,si,si,si,1,titulo,residencial,rocoso,15.00,8.30,si,156326,Barranca,40000,320.000000
5,6,125,llanura,horizontal,camino_medio_estrecho,si,si,si,0,titulo,residencial,rocoso,18.27,6.86,no,37121,Huaura,42000,336.000000
6,7,126,llanura,horizontal,camino_medio_estrecho,si,si,si,0,titulo,residencial,rocoso,12.00,10.50,no,37121,Huaura,38000,301.587302
7,8,130,llanura,horizontal,camino_medio_estrecho,si,si,si,1,titulo,residencial,rocoso,11.82,11.00,no,37121,Huaura,42000,323.076923
8,9,156,llanura,vertical,camino_medio_estrecho,si,si,si,1,titulo,residencial,rocoso,14.20,11.00,no,37121,Huaura,50000,320.512820
9,10,165,llanura,vertical,via_publica,si,si,si,2,titulo,residencial,rocoso,17.00,9.70,si,29671,Hualmay,54000,327.272727


##**Procesamiento de Datos**

In [ ]:
# Procesamiento de características categóricas con OneHotEncoder
encoder = OneHotEncoder()
ciudad_encoded = encoder.fit_transform(df[['ciudad']]).toarray()
zonificación_encoded = encoder.fit_transform(df[['zonificación']]).toarray()

# Escalamos las características numéricas para que estén en el mismo rango
scaler = MinMaxScaler()
numerical_features = scaler.fit_transform(df[['area', 'precio(mts2)']])

# Combinamos las características categóricas y numéricas
features = np.hstack((numerical_features, ciudad_encoded, zonificación_encoded))



In [ ]:
features[0]

#0,25: Este es el valor escalado para el 'área' del primer terreno.
#0.78: Este es el valor escalado para 'precio(mts2)' del primer terreno.
#1, 0, 0: representan la codificación única para 'ciudad'. Significa que el primer terreno pertenece a la primera ciudad en las categorías únicas identificadas por el codificador.
#1, 0: Representan la codificación one-hot para 'zonificación'. Significa que el primer terreno pertenece a la primera categoría de zonificación.


array([0.        , 0.73981083, 0.        , 1.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 1.        ])

In [ ]:
df

,id,area,topografia,forma,cam_conti,agua,luz,alcantarillado,dist_pist(Km),tip_doc,zonificación,tip_suelo,largo(mts),ancho(mts),vist_pano,población,ciudad,precio,precio(mts2)
0,1,108,llanura,horizontal,camino_medio_estrecho,si,si,si,1,titulo,residencial,rocoso,18.00,6.00,no,32312,Chancay,30000,277.777778
1,2,112,llanura,vertical,via_publica,si,si,si,2,titulo,residencial,rocoso,14.00,8.00,no,29671,Hualmay,30000,285.714286
2,3,120,llanura,vertical,camino_medio_estrecho,si,si,si,1,titulo,residencial,rocoso,15.00,8.00,si,156326,Barranca,35000,291.666667
3,4,121,llanura,horizontal,camino_medio_estrecho,si,si,si,0,titulo,residencial,rocoso,11.27,10.75,no,37121,Huaura,37000,305.785124
4,5,125,llanura,vertical,camino_medio_estrecho,si,si,si,1,titulo,residencial,rocoso,15.00,8.30,si,156326,Barranca,40000,320.000000
5,6,125,llanura,horizontal,camino_medio_estrecho,si,si,si,0,titulo,residencial,rocoso,18.27,6.86,no,37121,Huaura,42000,336.000000
6,7,126,llanura,horizontal,camino_medio_estrecho,si,si,si,0,titulo,residencial,rocoso,12.00,10.50,no,37121,Huaura,38000,301.587302
7,8,130,llanura,horizontal,camino_medio_estrecho,si,si,si,1,titulo,residencial,rocoso,11.82,11.00,no,37121,Huaura,42000,323.076923
8,9,156,llanura,vertical,camino_medio_estrecho,si,si,si,1,titulo,residencial,rocoso,14.20,11.00,no,37121,Huaura,50000,320.512820
9,10,165,llanura,vertical,via_publica,si,si,si,2,titulo,residencial,rocoso,17.00,9.70,si,29671,Hualmay,54000,327.272727


##**Calcua la similaridad de un caso**

In [ ]:
# Seleccionamos un terreno de referencia (por ejemplo, el id = 1)
terreno_referencia_idx = 0
terreno_referencia = features[terreno_referencia_idx]

# Calculamos la similitud del coseno entre el terreno de referencia y los demás terrenos
similitudes = cosine_similarity([terreno_referencia], features)[0]

# Creamos un DataFrame con los resultados ordenados por similitud
df['similitud'] = similitudes
recomendaciones = df.sort_values(by='similitud', ascending=False)

# Mostramos los terrenos más similares al terreno de referencia
print("Terrenos recomendados basados en la similitud:")
print(recomendaciones[['ciudad', 'area', 'zonificación', 'precio(mts2)', 'similitud']])


Terrenos recomendados basados en la similitud:
         ciudad  area zonificación  precio(mts2)  similitud
0       Chancay   108  residencial    277.777778   1.000000
21      Chancay   518  residencial    275.665399   0.998071
5        Huaura   125  residencial    336.000000   0.629359
11       Huaura   190  residencial    334.210526   0.628852
10       Huaura   183  residencial    327.868853   0.627104
13  Santa_maria   203  residencial    327.586207   0.627002
9       Hualmay   165  residencial    327.272727   0.626947
12      Vegueta   202  residencial    323.267327   0.625717
7        Huaura   130  residencial    323.076923   0.625711
8        Huaura   156  residencial    320.512820   0.624901
4      Barranca   125  residencial    320.000000   0.624750
14  Santa_maria   209  residencial    319.617225   0.624565
15      Vegueta   210  residencial    314.285714   0.622799
18  Santa_maria   356  residencial    308.988764   0.620598
3        Huaura   121  residencial    305.785124   0.

El resultado de este código será una tabla que muestra las tierras más similares a la tierra de referencia que seleccionó (la primera tierra en su conjunto de datos). La tabla está ordenada por similitud, por lo que las tierras más similares se encuentran en la parte superior.

Puntuación de similitud: la columna 'similitud' indica qué tan similar es cada terreno al terreno de referencia. Una puntuación de 1 significa que son idénticos y una puntuación más cercana a 0 significa que son menos similares.
Características: Las otras columnas ('ciudad', 'área', 'zonificación', 'precio(mts2)') muestran las características de cada terreno recomendado.
En esencia, el código busca tierras en su conjunto de datos que tengan características similares a las tierras de referencia, convirtiéndolas en posibles recomendaciones. Por ejemplo, si su terreno de referencia es un terreno grande y costoso en una ciudad específica con una determinada zonificación, las recomendaciones probablemente incluirán terrenos con atributos similares.

#**Ejemplo 2: Sistem de recomendaciòn con variable relevante**

In [ ]:
import pandas as pd
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder
import numpy as np

# Simulación de datos de terrenos optimizados
data=pd.read_csv('/content/drive/MyDrive/basedatos/BasedeDatosDefinitivaIDRELEVANTE.csv')
data

# Cargamos los datos en un DataFrame
df = pd.DataFrame(data)

# Procesamiento de características categóricas con OneHotEncoder
encoder = OneHotEncoder()
ciudad_encoded = encoder.fit_transform(df[['ciudad']]).toarray()
zonificación_encoded = encoder.fit_transform(df[['zonificación']]).toarray()

# Escalamos las características numéricas
scaler = MinMaxScaler()
numerical_features = scaler.fit_transform(df[['area', 'precio(mts2)']])

# Combinamos las características
features = np.hstack((numerical_features, ciudad_encoded, zonificación_encoded))

# Seleccionamos un terreno de referencia (id = 1)
terreno_referencia_idx = 0
terreno_referencia = features[terreno_referencia_idx]

# Calculamos la similitud del coseno
similitudes = cosine_similarity([terreno_referencia], features)[0]

# Creamos un DataFrame con los resultados ordenados por similitud
df['similitud'] = similitudes
recomendaciones = df.sort_values(by='similitud', ascending=False)

# --- Métricas de Evaluación ---
# Seleccionamos los primeros K terrenos como recomendados
K = 3
top_k = recomendaciones.iloc[:K]

# Precision@K
precision_at_k = top_k['relevante'].sum() / K

# Recall@K
total_relevantes = df['relevante'].sum()
recall_at_k = top_k['relevante'].sum() / total_relevantes if total_relevantes > 0 else 0

# F1-Score
f1_score = (2 * precision_at_k * recall_at_k) / (precision_at_k + recall_at_k) if (precision_at_k + recall_at_k) > 0 else 0

# Reporte
print("Reporte de rendimiento:")
print(f"Precision@{K}: {precision_at_k:.2f}")
print(f"Recall@{K}: {recall_at_k:.2f}")
print(f"F1-Score: {f1_score:.2f}")

# Recomendaciones ordenadas
print("\nTerrenos recomendados basados en la similitud:")
print(recomendaciones[['id', 'ciudad', 'area', 'zonificación', 'precio(mts2)', 'similitud', 'relevante']])


Reporte de rendimiento:
Precision@3: 0.67
Recall@3: 1.00
F1-Score: 0.80

Terrenos recomendados basados en la similitud:
    id       ciudad  area zonificación  precio(mts2)  similitud  relevante
0    1      Chancay   108  residencial    277.777778   1.000000          1
21  22      Chancay   518  residencial    275.665399   0.998071          1
5    6       Huaura   125  residencial    336.000000   0.629359          0
11  12       Huaura   190  residencial    334.210526   0.628852          0
10  11       Huaura   183  residencial    327.868853   0.627104          0
13  14  Santa_maria   203  residencial    327.586207   0.627002          0
9   10      Hualmay   165  residencial    327.272727   0.626947          0
12  13      Vegueta   202  residencial    323.267327   0.625717          0
7    8       Huaura   130  residencial    323.076923   0.625711          0
8    9       Huaura   156  residencial    320.512820   0.624901          0
4    5     Barranca   125  residencial    320.000000   

Informe de rendimiento:
Precisión@3: 0,67 significa que de las 3 principales recomendaciones, 2 fueron realmente relevantes (2/3 = 0,67).
Recordatorio@3: 0,20 indica que de todas las tierras relevantes en el conjunto de datos, solo el 20 % se incluyeron en las 3 principales recomendaciones.
Puntuación F1: 0,31 es una puntuación combinada que muestra un equilibrio moderado entre precisión y recordación.
Terrenos recomendados:
Esta tabla muestra las 3 principales tierras recomendadas (id 1, 4 y 3) con sus características y puntuaciones de similitud.
La columna relevante indica si una tierra recomendada fue realmente relevante (1 para relevante, 0 para no relevante).

#**Ejemplo 3: Sistema de recomendciòn considerando la relevancia como escala de likert**

In [ ]:
import pandas as pd
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder
import numpy as np

# Simulación de datos de terrenos con escala Likert de relevancia

data=pd.read_csv('/content/drive/MyDrive/basedatos/BasedeDatosDefinitivaIrelevancia_likert.csv')
data
# Cargamos los datos en un DataFrame
df = pd.DataFrame(data)

# Procesamiento de características categóricas con OneHotEncoder
encoder = OneHotEncoder()
ciudad_encoded = encoder.fit_transform(df[['ciudad']]).toarray()
tipo_encoded = encoder.fit_transform(df[['zonificación']]).toarray()

# Escalamos las características numéricas
scaler = MinMaxScaler()
numerical_features = scaler.fit_transform(df[['area', 'precio(mts2)']])

# Combinamos las características
features = np.hstack((numerical_features, ciudad_encoded, tipo_encoded))

# Seleccionamos un terreno de referencia (id = 1)
terreno_referencia_idx = 0
terreno_referencia = features[terreno_referencia_idx]

# Calculamos la similitud del coseno
similitudes = cosine_similarity([terreno_referencia], features)[0]

# Añadimos la similitud al DataFrame y ordenamos por relevancia
df['similitud'] = similitudes

# Aseguramos que la columna 'relevante' exista correctamente
relevancia_threshold = 4
df['relevante'] = (df['relevancia_likert'] >= relevancia_threshold).astype(int)

recomendaciones = df.sort_values(by='similitud', ascending=False)

# --- Métricas de Evaluación ---
# Definimos un umbral para considerar relevante (Likert >= 4)
#relevancia_threshold = 4

# Aseguramos que la columna 'relevante' exista correctamente
#df['relevante'] = (df['relevancia_likert'] >= relevancia_threshold).astype(int)

# Seleccionamos los primeros K terrenos como recomendados
K = 3
top_k = recomendaciones.iloc[:K]

# Precision@K
precision_at_k = top_k['relevante'].sum() / K

# Recall@K
total_relevantes = df['relevante'].sum()
recall_at_k = top_k['relevante'].sum() / total_relevantes if total_relevantes > 0 else 0

# F1-Score
f1_score = (2 * precision_at_k * recall_at_k) / (precision_at_k + recall_at_k) if (precision_at_k + recall_at_k) > 0 else 0

# Weighted Precision: ponderamos por la escala Likert
weighted_precision = top_k.apply(lambda x: x['relevancia_likert'] if x['relevante'] == 1 else 0, axis=1).sum() / (K * 5)

# Reporte
print("Reporte de rendimiento:")
print(f"Precision@{K}: {precision_at_k:.2f}")
print(f"Recall@{K}: {recall_at_k:.2f}")
print(f"F1-Score: {f1_score:.2f}")
print(f"Weighted Precision: {weighted_precision:.2f}")

# Recomendaciones ordenadas
print("\nTerrenos recomendados basados en la similitud:")
print(recomendaciones[['id', 'ciudad', 'area', 'zonificación', 'precio(mts2)', 'similitud', 'relevancia_likert', 'relevante']])


Reporte de rendimiento:
Precision@3: 1.00
Recall@3: 1.00
F1-Score: 1.00
Weighted Precision: 0.93

Terrenos recomendados basados en la similitud:
    id       ciudad  area zonificación  precio(mts2)  similitud  \
0    1      Chancay   108  residencial    277.777778   1.000000   
21  22      Chancay   518  residencial    275.665399   0.998071   
5    6       Huaura   125  residencial    336.000000   0.629359   
11  12       Huaura   190  residencial    334.210526   0.628852   
10  11       Huaura   183  residencial    327.868853   0.627104   
13  14  Santa_maria   203  residencial    327.586207   0.627002   
9   10      Hualmay   165  residencial    327.272727   0.626947   
12  13      Vegueta   202  residencial    323.267327   0.625717   
7    8       Huaura   130  residencial    323.076923   0.625711   
8    9       Huaura   156  residencial    320.512820   0.624901   
4    5     Barranca   125  residencial    320.000000   0.624750   
14  15  Santa_maria   209  residencial    319.61722

Precision@3: 1.00: Esto significa que las 3 recomendaciones principales (K=3) son relevantes para el usuario. En otras palabras, todas las recomendaciones proporcionadas por el sistema son acertadas.

Recall@3: 1.00: Esto significa que todos los terrenos relevantes en el conjunto de datos fueron incluidos en las 3 principales recomendaciones. El sistema no ha pasado por alto ningún terreno relevante.

F1-Score: 1.00: El F1-Score es una medida que combina la precisión y el recall. Un valor de 1.00 indica un rendimiento perfecto, lo que significa que el sistema tiene una precisión y un recall perfectos.

Weighted Precision: 0.93: Esta métrica es similar a la precisión, pero ponderada por la escala Likert, dando más peso a los terrenos con mayor relevancia. Un valor de 0.93 indica que las recomendaciones tienen una alta relevancia, aunque no perfecta, teniendo en cuenta la escala Likert.